src/java.base/share/classes/java/util/concurrent/Joiners.java

```diff
@@ -0,0 +1,220 @@
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package java.util.concurrent;
+
+import java.lang.invoke.MethodHandles;
+import java.lang.invoke.VarHandle;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.Objects;
+import java.util.concurrent.StructuredTaskScope.Joiner;
+import java.util.concurrent.StructuredTaskScope.Subtask;
+import java.util.function.Predicate;
+import java.util.stream.Stream;
+import jdk.internal.invoke.MhUtil;
+
+/**
+ * Built-in StructuredTaskScope.Joiner implementations.
+ */
+class Joiners {
+    private Joiners() { }
+
+    /**
+     * Throws IllegalArgumentException if the subtask is not in the UNAVAILABLE state.
+     */
+    private static void ensureUnavailable(Subtask<?> subtask) {
+        if (subtask.state() != Subtask.State.UNAVAILABLE) {
+            throw new IllegalArgumentException("Subtask not in UNAVAILABLE state");
+        }
+    }
+
+    /**
+     * Throws IllegalArgumentException if the subtask has not completed.
+     */
+    private static Subtask.State ensureCompleted(Subtask<?> subtask) {
+        Subtask.State state = subtask.state();
+        if (state == Subtask.State.UNAVAILABLE) {
+            throw new IllegalArgumentException("Subtask has not completed");
+        }
+        return state;
+    }
+
+    /**
+     * A joiner that returns a stream of all subtasks when all subtasks complete
+     * successfully. Cancels the scope if any subtask fails.
+     */
+    static final class AllSuccessful<T> implements Joiner<T, Stream<Subtask<T>>> {
+        private static final VarHandle FIRST_EXCEPTION =
+                MhUtil.findVarHandle(MethodHandles.lookup(), "firstException", Throwable.class);
+
+        // list of forked subtasks, only accessed by owner thread
+        private final List<Subtask<T>> subtasks = new ArrayList<>();
+
+        private volatile Throwable firstException;
+
+        @Override
+        public boolean onFork(Subtask<? extends T> subtask) {
+            ensureUnavailable(subtask);
+            @SuppressWarnings("unchecked")
+            var s = (Subtask<T>) subtask;
+            subtasks.add(s);
+            return false;
+        }
+
+        @Override
+        public boolean onComplete(Subtask<? extends T> subtask) {
+            Subtask.State state = ensureCompleted(subtask);
+            return (state == Subtask.State.FAILED)
+                    && (firstException == null)
+                    && FIRST_EXCEPTION.compareAndSet(this, null, subtask.exception());
+        }
+
+        @Override
+        public Stream<Subtask<T>> result() throws Throwable {
+            Throwable ex = firstException;
+            if (ex != null) {
+                throw ex;
+            } else {
+                return subtasks.stream();
+            }
```

Comment:
```suggestion
            return subtasks.stream();
```
The else here is redundant